### PR TITLE
Eliminate FromJavaObject from TASTy of Java sources

### LINKED ISSUE #19246: Ensure consistent JAVAattr tasty output
## Compiler version

3.4.0-RC1-bin-SNAPSHOT

## Minimized code

Before we freeze 28.4.0 TASTy, we should ensure that JAVAattr output is consistent.

Currently `<FromJavaObject>` does appear in TASTy generated for Java sources. We should ensure that it is consistently applied (i.e. there is not a mix of `java.lang.Object` and `<FromJavaObject>`)

### PR DESCRIPTION
**Pull request by bishabosha**
Do not pickle type references to `FromJavaObject` for Java sources, instead insert them at unpickling time. This ensures consistency of `IDENTtpt` (where `Object` is given an explicit type) and `SELECTtpt` where the type has to be resolved from just the name `Object`.

modify `-Ytest-pickler` to check for the presence of `FromJavaObject`

Also we compare unpicking Java signatures from TASTy to unpicking from class files to ensure consistency

fixes #19246

### CODE REVIEW

**Comment by bishabosha** [comment]
We are also considering dropping the "fake" `public Foo(scala.Unit x$1) {}` constructor added to all java classes by the parser

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 1]
```diff
@@ -188,19 +191,19 @@ class TreePickler(pickler: TastyPickler, attributes: Attributes) {
...
-        }
-        else {
+        else if isJavaPickle && sym == defn.FromJavaObjectSymbol then
+          pickleType(defn.ObjectType) // when unpickling Java TASTy, replace by <FromJavaObject>
+        else
           writeByte(if (tpe.isType) TYPEREF else TERMREF)
           pickleNameAndSig(sym.name, tpe.signature, sym.targetName)
-          pickleType(tpe.prefix)
-        }
+            pickleType(tpe.prefix)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:206]: Something's off with the indentation here.

**Review by sjrd — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala (hunk 1 above)
Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala]: thanks!

**Comment by bishabosha** [comment]
@sjrd no idea why but the spec keeps failing to build due to ruby errors unrelated to this PR

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 2]
```diff
@@ -1064,37 +1064,37 @@ class TreeUnpickler(reader: TastyReader,
...
+              lazy val derivedTparamSyms: List[TypeSymbol] = tparams.map: tdef =>
                 val completer = new LazyType {
                   def complete(denot: SymDenotation)(using Context) =
-                    denot.info = tdef.symbol.asType.info.subst(tdefRefs, tparamSyms.map(_.typeRef))
+                    denot.info = tdef.symbol.asType.info.subst(tparamRefs, derivedTparamRefs)
                 }
-                newSymbol(sym, tdef.name, pflags, completer, coord = cls.coord)
-              val paramSym =
+                newSymbol(sym, tdef.name, Flags.JavaDefined | Flags.Param, completer, coord = cls.coord)
+              lazy val derivedTparamRefs: List[Type] = derivedTparamSyms.map(_.typeRef)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1084]: This is only used once. It can be a `def`.

Comment (sjrd) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1084]: Also, maybe inline it back into `def complete`, like it was before?

**Review by sjrd — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala (hunk 2 above)
Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1084]: The point is that each type parameter has a new completer, so this is referenced arbitrary times

**Comment by bishabosha** [comment]
ignoring the weird spec CI error

**Pull request merged**